### PR TITLE
[ONNX] Fix scalar type cast for comparison ops

### PR DESCRIPTION
**Pull request by BowenBao**
Always promote type casts for comparison operators, regardless if the input is tensor or scalar. Unlike arithmetic operators, where scalars are implicitly cast to the same type as tensors.

### CODE REVIEW

**Comment by BowenBao** [comment]
cc @jacky82226

**Comment by BowenBao** [comment]
Test `TestONNXRuntime_opset11.test_mask_rcnn` failing with 
```
E       onnxruntime.capi.onnxruntime_pybind11_state.NotImplemented: [ONNXRuntimeError] : 9 : NOT_IMPLEMENTED : Could not find an implementation for the node Greater_1611:Greater(9)
```
This PR is currently blocked due to that Greater[double] is not supported in onnxruntime.

**Review by houseroad — COMMENTED** [review]
Any update on this PR?

**Comment by BowenBao** [comment]
> 
> 
> Any update on this PR?

Will submit a fix to ONNXRuntime to unblock this.

**Review by bzinodev — APPROVED** [review]

**Review by bzinodev — CHANGES_REQUESTED** [review]
Need Microsoft approvers

File: torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp [hunk 1]
```diff
@@ -121,30 +121,41 @@ static c10::optional<c10::ScalarType> InferExpectedScalarType(const Node* n) {
...
+              typesFromTensors.end(),
+              [&st](const c10::ScalarType& type) { return type != st; })) {
+        std::cerr
+            << "Warning: ONNX Scalar Type Analysis - Scalar types mismatch for tensor inputs of operator "
+            << n->kind().toDisplayString()
+            << ". Please report a bug to PyTorch. "
+            << "The scalar type " << c10::toString(*st)
+            << " of the first tensor is chosen." << std::endl;
+      }
+    } else {
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp:155]: Is this different than the first condition in line 133?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -2699,7 +2699,7 @@ def forward(self, x):
 
         class ComparisonModel(torch.nn.Module):
             def forward(self, x, y):
-                return x.ge(0.5) & y.le(2)
+                return x.lt(1.5) & y.le(2) & x.le(1)
```
Comment (neginraoof) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:2702]: Can you add test cases for when all inputs are tensors, or all scalars?
Does this cover the later case?

**Review by neginraoof — COMMENTED** [review]
Thanks. Had a couple comments.

**Comment by BowenBao** [comment]
The CI failure needs a fix from ONNXRuntime, which is already in master and will be reflected in 7.29's nightly build.

File: torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp (hunk 1 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp:155]: yes, though it also derives types from scalar inputs, the priority is much lower, as tensor input type and output type are both higher priority.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 2 above)
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: added

**Review by neginraoof — APPROVED** [review]
LGTM. Thanks!

**Review by bzinodev — APPROVED** [review]

**Comment by BowenBao** [comment]
@bzinodev Can you import this PR for merge? Thanks.

**Pull request closed without merging**